taichi/transforms/constant_fold.cpp

```diff
@@ -1,9 +1,72 @@
 #include "taichi/ir/ir.h"
 #include <deque>
 #include <set>
+#include <cmath>
 
 TLANG_NAMESPACE_BEGIN
 
+bool TypedConstant::from_unary_op(UnaryOpType op, const TypedConstant &rhs)
+{
+#define PER_OP(op, o) \
+    case UnaryOpType::op: this->val_i32 = o(rhs.val_i32); break;
+  switch (op) {
+  PER_OP(neg, -)
+  PER_OP(sqrt, std::sqrt)
+  PER_OP(log, std::log)
+  PER_OP(exp, std::exp)
+  PER_OP(abs, std::abs)
+  PER_OP(sin, std::sin)
+  PER_OP(cos, std::cos)
+  PER_OP(tan, std::tan)
+  PER_OP(asin, std::asin)
+  PER_OP(acos, std::acos)
+  PER_OP(tanh, std::tanh)
+  PER_OP(rsqrt, 1 / std::sqrt)
```

Comment:
These are incongruent with `this->val_i32 = o(rhs.val_i32);`... We need to be cautious about the return type of unary operators.